src/resolution/instantiation.ts

```diff
@@ -33,17 +13,39 @@ function _createInstance<T>(
     let result: T | Promise<T>;
 
     if (childRequests.length > 0) {
-        const constructorInjections = _getConstructionInjections(childRequests, resolveRequest);
-
-        if (constructorInjections.some(isPromise)) {
-            result = _createInstanceWithConstructorInjectionsAsync(
-                constructorInjections,
-                constr,
-                childRequests,
-                resolveRequest
-            );
-        } else {
-            result = _createInstanceWithConstructorInjections(constructorInjections, constr,childRequests, resolveRequest);
+        let isAsync = false
+        const constructorInjections: unknown[] = []
+        const propertyRequests: interfaces.Request[] = []
+        const propertyInjections: unknown[] = []
+        for(const childRequest of childRequests){
+            let injection:unknown
+            const target = childRequest.target
+            const targetType = target.type
+            if(targetType === TargetTypeEnum.ConstructorArgument){
+                injection = resolveRequest(childRequest)
+                constructorInjections.push(injection)
+            }else{
+                propertyRequests.push(childRequest)
+                injection = resolveRequest(childRequest)
+                propertyInjections.push(injection)
+            }
+            if(!isAsync){
+                if(Array.isArray(injection)){
+                    for(const arrayInjection of injection){
+                        if(isPromise(arrayInjection)){
+                            isAsync = true
+                            break
+                        }
+                    }
+                }else if(isPromise(injection)){
+                    isAsync = true
+                }
```

Comment:
I think we should try to extract this logic to a method, otherwise the function is too big

Reply:
Please feel free - it did pass code climate though.  There are other code climate warnings if you are interested !